File: Breadth-first_search(BFS)/python/01_숨바꼭질.py
# fast campus 강의
# https://www.acmicpc.net/problem/1697
# 1

from collections import deque

n, k = map(int, input().split())
visited = [0] * 100001

def bfs():
    q = deque([n])
    while q:
        x = q.popleft()
        if x == k:
            return visited[x]
        for next in (x - 1, x + 1, x * 2):
            if 0 <= next < 100001 and visited[next] == 0:
                visited[next] = visited[x] + 1
                q.append(next)

print(bfs())

File: Breadth-first_search(BFS)/c++/01_숨바꼭질.cpp
// fast campus 강의
// https://www.acmicpc.net/problem/1697
// 1

#include <iostream>
#include <queue>
#define MAX 100001

using namespace std;

int bfs(int n, int k)
{
    queue<int> q;
    int cost[MAX] = {0, };
    q.push(n);

    while (!q.empty()) {
        int now = q.front();
        q.pop();

        if (now == k)
            return cost[now];
        
        int next = now - 1;
        if (0 <= next && next < MAX && !cost[next]) {
            cost[next] = cost[now] + 1;
            q.push(next);
        }

        next = now + 1;
        if (0 <= next && next < MAX && !cost[next]) {
            cost[next] = cost[now] + 1;
            q.push(next);
        }

        next = now * 2;
        if (0 <= next && next < MAX && !cost[next]) {
            cost[next] = cost[now] + 1;
            q.push(next);
        }
    }
}

int main()
{
    int N, K;
    cin >> N >> K;

    cout << bfs(N, K);
}